vllm/logger.py

```diff
@@ -1,72 +1,93 @@
-
-    def __init__(self, fmt, datefmt=None):
-        logging.Formatter.__init__(self, fmt, datefmt)
-
-    def format(self, record):
-        msg = logging.Formatter.format(self, record)
-        if record.message != "":
-            parts = msg.split(record.message)
-            msg = msg.replace("\n", "\r\n" + parts[0])
-        return msg
-
-
-_root_logger = logging.getLogger("vllm")
-_default_handler: Optional[logging.Handler] = None
-
-
-def _setup_logger():
-    _root_logger.setLevel(logging.DEBUG)
-    global _default_handler
-    if _default_handler is None:
-        _default_handler = logging.StreamHandler(sys.stdout)
-        _default_handler.flush = sys.stdout.flush  # type: ignore
-        _default_handler.setLevel(logging.INFO)
-        _root_logger.addHandler(_default_handler)
-    fmt = NewLineFormatter(_FORMAT, datefmt=_DATE_FORMAT)
-    _default_handler.setFormatter(fmt)
-    # Setting this will avoid the message
-    # being propagated to the parent logger.
-    _root_logger.propagate = False
+DEFAULT_LOGGING_CONFIG = {
+    "formatters": {
+        "vllm": {
+            "class": "vllm.logging.NewLineFormatter",
+            "datefmt": _DATE_FORMAT,
+            "format": _FORMAT,
+        },
+    },
+    "handlers": {
+        "vllm": {
+            "class": "logging.StreamHandler",
+            "formatter": "vllm",
+            "level": "INFO",
+            "stream": "ext://sys.stdout",
+        },
+    },
+    "loggers": {
+        "vllm": {
+            "handlers": ["vllm"],
+            "level": "DEBUG",
+            "propagate": False,
+        },
+    },
+    "version": 1,
+}
+
+
+def _configure_vllm_root_logger() -> None:
+    if VLLM_CONFIGURE_LOGGING:
+        logging_config: Dict = DEFAULT_LOGGING_CONFIG
+
+    if VLLM_LOGGING_CONFIG_PATH:
+        if not path.exists(VLLM_LOGGING_CONFIG_PATH):
+            raise RuntimeError(
+                "Could not load logging config. File does not exist:"
+                f" {VLLM_LOGGING_CONFIG_PATH}")
+        with open(VLLM_LOGGING_CONFIG_PATH, encoding="utf-8",
+                  mode="r") as file:
+            custom_config = json.loads(file.read())
+
+        if not isinstance(custom_config, dict):
+            raise ValueError("Invalid logging config. Expected Dict, got"
+                             f" {type(custom_config).__name__}.")
+        logging_config = custom_config
+
+    if logging_config:
+        dictConfig(logging_config)
+
+
+def _configure_vllm_logger(logger: Logger) -> None:
+    # Use the same settings as for root logger
+    _root_logger = logging.getLogger("vllm")
+    default_log_level = os.getenv("LOG_LEVEL", _root_logger.level)
+    logger.setLevel(default_log_level)
+    for handler in _root_logger.handlers:
+        logger.addHandler(handler)
+    logger.propagate = False
+
+
+def init_logger(name: str) -> Logger:
+    logger_is_new = name not in logging.Logger.manager.loggerDict
```

Comment:
It's unclear to me if `logging.Logger.manager.loggerDict` is part of the official API, but this is the best way I've found to tell if a logger has already been created/configured.

Reply:
I think it is a private API, though it is unlikely going to be changed.

What's happening if we don't check this at all? Just reconfiguring logger which seems okay to me?

Reply:
The goal here is to avoid configuring a vllm logger that's been explicitly configured via the dict given to `loggerDict`.

More concretely, if the logging config file included config for `vllm.other`, this logic aims to avoid overwriting that config with the config that would be applied to loggers not configured with the logging config file.

If we move in the direction of leveraging `propagate=True`, this edge case goes away and that's the way I'm leaning. I feel like the whole thing is more easily explained/understood with `propagate=True`.

Reply:
This code was removed when making updates to better rely on logger propagation.